### ISSUE TITLE
dbt-utils "star" returns all columns even if all columns are listed in except list

### ISSUE THREAD

**Issue by KennyShaevelA8** [dbt-labs/dbt-utils#681]
### Describe the bug
When using the `star` dbutils function, if you (inadvertently) include *every* column of a table in the `except` list, star.sql returns `*`. This leads to unexpected behavior where the query attempts to return all columns. (I would expect star.sql to return an empty string in this situation.)

In the web app, the SQL compiles fine, adding the * in where expected. Bizarrely, it even previews just fine, as if the * were not there. The issue only arises when doing `dbt run`. 

### Steps to reproduce
A simplified example of where I discovered it. 

```
with salesorderdetail as (
  select * from {{ ref( 'fact_sales_order_detail') }}
),

dim_product as (
  select 
    column_a,
    column_b,
    column_c
  from {{ ref('dim_product') }}
),

select
    sod.*,
    {{ dbt_utils.star(from=ref('dim_product'), relation_alias='pr', except=["column_a", "column_b", "column_c"]) }},
from salesorderdetail sod
    left join dim_product pr
        on pr.column_a = sod.product_id
```

### Expected results
I suppose I would expect a compile error to be thrown, spurred by an error return from `star`. 

### Actual results
The query compiles just fine, adding a `*` in where the `star` function was called. No error is even seen when previewing the query, although you can see the `*` in the compiled query. Specifically, the above query would compile to:

```
with salesorderdetail as (

  select * from [long string pointing to fact_sales_order_detail]
),

dim_product as (
  
  select 
    column_a,
    column_b,
    column_c
  from [long string pointing to dim_product]
),

select
    sod.*,
    *,
from salesorderdetail sod
    left join dim_product pr
        on pr.column_a = sod.product_id
```

This attempts to load all the columns of `sod` twice, leading to a compile error due to having multiple columns with identical names. 

### System information
**The contents of your `packages.yml` file:**
```
packages:
  - package: dbt-labs/dbt_utils
    version: 0.9.1
```

**Which database are you using dbt with?**
- [ ] postgres
- [ ] redshift
- [ ] bigquery
- [x] snowflake
- [ ] other (specify: ____________)

**The output of `dbt --version`:**
1.2.0

### Are you interested in contributing the fix?
Sure but not 100% sure how. There needs to be some handling to determine if all columns appear in the `except` list, either in `star` or in `get_filtered_columns_in_relation` I believe. Perhaps a check as simple as if the `except` list length is equal to (or greater) the table's column list length.

**Comment by joellabes** [comment]
Hey @KennyShaevelA8, I agree that this is a bit of a surprise! 

Its origin can be traced back with three whys: 

At the most obvious level, it happens because of this: 
https://github.com/dbt-labs/dbt-utils/blob/ba7ee523b0833d1564721f92bbdfe3cf82a5e62b/macros/sql/star.sql#L16-L18

But why does that happen? It's because we start with an empty list, and add desired columns to it: 
https://github.com/dbt-labs/dbt-utils/blob/ba7ee523b0833d1564721f92bbdfe3cf82a5e62b/macros/sql/get_filtered_columns_in_relation.sql#L18-L20

And why does _that_ happen? Best I can tell, this dates back to https://github.com/dbt-labs/dbt-utils/pull/561 which is dealing with the fact that SQLFluff doesn't like select statements without any columns specified. 

What to do about it? I'm not sure! The reason specified for the change in #561 was that a model doesn't exist at the time that SQLFluff runs.

Perhaps instead of returning `'*'`, it should return a static string like `'no columns provided to star() macro'` - valid SQL, but clearly wrong for analytics purposes. @swanjson what do you think?

**Comment by swanjson** [comment]
@joellabes without looking at the code again, would there be a way to determine the difference between an empty string created by "except" columns and an empty string due to the table not existing? The default * is what allows sqlfluff to run if the model hasn't been created before. It seems the rarer case to have all the columns of model in the except clause than linting on a file that has not been built yet.

**Comment by joellabes** [comment]
I had a quick look at identifying whether it's because the model hasn't been built yet, but it's given me trouble in the past so I wouldn't want to rush back into it. 

>The default * is what allows sqlfluff to run if the model hasn't been created before

I know, but it doesn't have to be a star, right? It just has to be any valid SQL syntax, which is why I thought a static string might be a better compromise - the CI robots don't care, and it's more obvious to the humans what happened. Right now the result of the macro is the exact opposite of what was requested which is sorta weird.

**Comment by joellabes** [comment]
Coming back to this: [I asked in the sqlfluff channel](https://getdbt.slack.com/archives/CV2KM39KR/p1663648576690009?thread_ts=1654191972.297119&cid=CV2KM39KR) and got a "sounds okay, I think" from @barrywhart which I'm taking as a full endorsement and moving forward 🚀

**Comment by joellabes** [comment]
Closed by #732
